test/boost/sstable_datafile_test.cc

```diff
@@ -1036,6 +1036,76 @@ static flat_mutation_reader make_normalizing_sstable_reader(shared_sstable sst,
     return make_normalizing_sstable_reader(sst, s, query::full_partition_range);
 }
 
+SEASTAR_TEST_CASE(autocompaction_control_test) {
+    return test_env::do_with_async([] (test_env& env) {
+        storage_service_for_tests ssft;
+        BOOST_REQUIRE(smp::count == 1);
+
+        auto cm = make_lw_shared<compaction_manager>();
+        cm->start();
+
+        schema_builder builder(make_lw_shared(schema({}, some_keyspace, some_column_family,
+            {{"p1", utf8_type}}, {{"c1", utf8_type}}, {{"r1", utf8_type}}, {}, utf8_type)));
+        builder.set_min_compaction_threshold(2);
+        auto s = builder.build(schema_builder::compact_storage::yes);
+
+        auto tmp = tmpdir();
+        column_family::config cfg = column_family_test_config();
+        cfg.datadir = tmp.path().string();
+        cfg.enable_commitlog = false;
+        cfg.enable_incremental_backups = false;
+        auto cl_stats = make_lw_shared<cell_locker_stats>();
+        auto tracker = make_lw_shared<cache_tracker>();
+        auto cf = make_lw_shared<column_family>(s, cfg, column_family::no_commitlog(), *cm, *cl_stats, *tracker);
+        cf->start();
+        cf->mark_ready_for_writes();
+        cf->set_compaction_strategy(sstables::compaction_strategy_type::size_tiered);
+
+        // create and add sstables
+        auto sst_gen = [&env, s, &tmp, gen = make_lw_shared<unsigned>(1)] () mutable {
+            return env.make_sstable(s, tmp.path().string(), (*gen)++, la, big);
+        };
+
+        auto make_insert = [&] (partition_key key) {
+            mutation m(s, key);
+            m.set_clustered_cell(clustering_key::from_exploded(*s, {to_bytes("c1")}), bytes("r1"), data_value(int32_t(1)), 1);
+            return m;
+        };
+
+        auto alpha = partition_key::from_exploded(*s, {to_bytes("alpha")});
```

Comment:
this is the reason you need smp ==1 in the first place, and we currently have a way to work around it.

There is a function in `test/lib/sstable_utils.cc` called `token_generation_shard`

Please see an example of its usage in the resharding test:

```
       auto key_token_pair = token_generation_for_shard(keys_per_shard, i, cfg->murmur3_partitioner_ignore_msb_bits());
            BOOST_REQUIRE(key_token_pair.size() == keys_per_shard);
```

It is important to pass the correct msb. There is another function btw that is `token_generation_for_shard` it is probably broken as it doesn't pass the right msb. That one is actually used in the sstable_datafile_test which is why we don't notice, but don't use that.

Reply:
Sadly saying, this:
```
    auto key_and_token_pair = token_generation_for_current_shard(50);
    auto min_key = key_and_token_pair[0].first;
    auto max_key = key_and_token_pair[key_and_token_pair.size()-1].first;
    auto max_sstable_size_in_mb = 1;
    auto max_sstable_size = max_sstable_size_in_mb*1024*1024;

    // Creating two sstables which key range overlap.
    add_sstable_for_leveled_test(env, cf, /*gen*/1, max_sstable_size, /*level*/0, min_key, max_key);
    BOOST_REQUIRE(cf->get_sstables()->size() == 1);

    add_sstable_for_leveled_test(env, cf, /*gen*/2, max_sstable_size, /*level*/0, key_and_token_pair[1].first, max_key);
    BOOST_REQUIRE(cf->get_sstables()->size() == 2);
```
did not help. table compaction fails with an assertion `sstables/sstables.cc:2538: uint64_t sstables::sstable::bytes_on_disk() const: Assertion `_bytes_on_disk > 0' failed.`
I will have to add ss tables manually with mutations.

Reply:
```
auto alpha = partition_key::from_exploded(*s, {to_bytes("alpha")});
```

> this is the reason you need smp ==1 in the first place, and we currently have a way to work around it.

@glommer Why? I don't see the connection between `smp` number and what I put into sstable. As far as I see `partition_key::from_exploded` only serializes a key. I don't know why it is called `exploded` btw. My guess is that you meant that maybe I need properly generated tokens for building overlapping sstables? It seems like everything works without it. However, I am not sure if it is right.
 
> There is a function in `test/lib/sstable_utils.cc` called `token_generation_shard`

No test uses this function. However, almost all of them use an alias to it `token_generation_for_current_shard(tokens_to_generate) = token_generation_for_shard(tokens_to_generate, this_shard_id(), unsigned ignore_msb = 0, unsigned smp_count = smp::count);` either for generating overlapping sstables or for uses tokens instead of the keys. The rest half of the tests just uses `from_explode` to generate keys for mutations.

Basically here what they do:

```
        auto make_insert = [&] (partition_key key) {
            mutation m(s, key);
            // TODO: replace this with token_generation_shard
            m.set_clustered_cell(clustering_key::from_exploded(*s, {to_bytes("c1")}), bytes("r1"), data_value(int32_t(1)), 1);
            return m;
        };

        // TODO: replace this with token_generation_shard
        auto alpha = partition_key::from_exploded(*s, {to_bytes("alpha")});

        column_family_test(cf).add_sstable(make_sstable_containing(sst_gen, {make_insert(alpha)}));
        column_family_test(cf).add_sstable(make_sstable_containing(sst_gen, {make_insert(alpha)}));
```

or 

```
 auto tokens = token_generation_for_current_shard(4);
        auto make_insert = [&] (auto p) {
            auto key = partition_key::from_exploded(*s, {to_bytes(p.first)});
            mutation m(s, key);
            m.set_clustered_cell(clustering_key::make_empty(), bytes("value"), data_value(int32_t(1)), 1 /* ts */);
            return m;
        };
        auto mut1 = make_insert(tokens[0]);
        auto mut2 = make_insert(tokens[1]);
        auto mut3 = make_insert(tokens[2]);
        auto mut4 = make_insert(tokens[3]);
        std::vector<shared_sstable> ssts = {
                make_sstable_containing(sst_gen, {mut1, mut2}),
                make_sstable_containing(sst_gen, {mut3, mut4})
        };
```

from my perspective of very limited knowledge of tables internals and my goals, it looks almost the same. So the question remains, how SMP is connected to the from_exploded.

Reply:
> this is the reason you need smp ==1 in the first place, and we currently have a way to work around it.

My guess is it is because we want to put proper tokens into the sstables relative to shard?

Reply:
well yes, but there is now a much better way to do it - as you did. This can be resolved.